Feature/14/store screen

brandonlee503 opened the pull request:
Main store screen implemented. Some of the commits are irrelevant - pulled from tools branch 😆

willmichael committed:
Add Activity View Cells

willmichael committed:
Add sub views and controllers for ActivityView

willmichael committed:
Add healthKit capabailities, add basic healthKit Manager, add info.plist descriptions

willmichael committed:
Add Query function into HealthKitManager

willmichael committed:
Add basic MapKit functionality to MapView

brandonlee503 committed:
Merge branch 'feature/15/tools-screen' of https://github.com/iOS-Capstone/C7FIT into dev

brandonlee503 committed:
Remove Main.Storyboard :)

brandonlee503 committed:
Update NSPhotoLibraryUsageDescription

willmichael committed:
Add start/stop run tracking with time, distance, and pace

willmichael committed:
Add Save runs

brandonlee503 committed:
Add launch icons

brandonlee503 committed:
Remove launch screen storyboard :)

brandonlee503 committed:
Add app icon

willmichael committed:
Add detail pane to display end run statistics

brandonlee503 committed:
Add OAuthSwift Pod for OAuth 2.0

brandonlee503 committed:
Move view folder structure

brandonlee503 committed:
Remove template store view

brandonlee503 committed:
Delete template profile view

brandonlee503 committed:
Set up collection view controller template

brandonlee503 committed:
Implement initial eBayDataManager

brandonlee503 committed:
Remove OAuthSwift

brandonlee503 committed:
Implement eBay getItem

brandonlee503 committed:
Switch to production API :D

willmichael committed:
Add run detail pane, highlights the user's run path

willmichael committed:
Add saving user runs to firebase, update FirebaseDataManager and runData models to accomodate

brandonlee503 committed:
Merge tools screen

brandonlee503 committed:
Tidy up after merge

brandonlee503 committed:
Add searchItem query

brandonlee503 committed:
Limit search to 10 items

brandonlee503 committed:
Build initial collection view UI template

brandonlee503 committed:
Add eBayItem model

brandonlee503 committed:
Get store screen initially displaying JSON data

brandonlee503 committed:
Separated CategoryCell view and controller

brandonlee503 committed:
Restructure project files

brandonlee503 committed:
Get images loading on store screen

brandonlee503 committed:
Make item cells UI prettier

brandonlee503 committed:
Use UIImageView extension in profile screen

brandonlee503 commented:
From the travis log, the build fails in `eBayDataManager.getOAuth2Token()` from the following:
```
UI Testing Failure - com.ClubSevenFitnessTest2.C7FIT crashed in C7FIT.eBayDataManager.getOAuth2Token (completion : (Swift.Optional<Swift.String>) -> ()) -> ()
```

It's likely that this is due to the credentials being kept hidden from the repo.. Need to find a way to make this pass while keeping variables hidden.

brandonlee503 committed:
Improve some existing implementations

brandonlee503 committed:
Move UIImageView extension into its own subclass with caching

brandonlee503 committed:
Reformat eBayDataManager to exclude fetching token - single responsibility principle

brandonlee503 commented:
Travis error fixed with https://github.com/iOS-Capstone/C7FIT/pull/47/commits/e96c02043f8bc0ac896da802cdc235eea48638dd

splons commented on C7FIT/C7FUIImageView.swift:
```diff
@@ -0,0 +1,45 @@
+            URLSession.shared.dataTask(with: url) { data, response, error in
+                guard let data = data, error == nil else { return }
+                DispatchQueue.main.async {
+                    let imageToCache = UIImage(data: data)
```
seems like a good place for an `if let` to guard against bad data (and a potential crash)

brandonlee503 replied:
Wouldn't the guard on line 36 be sufficient?
`guard let data = data, error == nil else { return }`

splons replied:
Certainly is a good start. But what if the data does not make a valid image? Then imageToCache will be nil and the force unwrap will be a crash.

splons commented on C7FIT/CategoryCell.swift:
```diff
@@ -0,0 +1,77 @@
+        itemCollectionView.backgroundColor = .blue
+        addSubview(itemCollectionView)
+        
+        dividerLine.backgroundColor = .yellow//UIColor(white: 0.5, alpha: 0.5)
```
spiffy color scheme!

splons commented on C7FIT/HealthKitManager.swift:
```diff
@@ -0,0 +1,91 @@
+                return
+            }
+        
+            let mostRecent = results?.first as? HKQuantitySample
```
use `if let` here instead of explicit test for nil and force upwrap

splons commented on C7FIT/HealthKitViewController.swift:
```diff
@@ -0,0 +1,74 @@
+        let sampleType = HKSampleType.quantityType(forIdentifier: HKQuantityTypeIdentifier.bodyMass)
+        self.healthKitManager.queryUserData(sampleType: sampleType!, completion : { (mostRecentWeight, error) -> Void in
+            if (error != nil){
+                print("Error")
```
even better would be `print(error)` so you log the actual error contents

splons commented on C7FIT/HealthKitViewController.swift:
```diff
@@ -0,0 +1,74 @@
+            //placed in different thread
+            DispatchQueue.main.async() {
+                //update the text async
+                self.healthView.weightLabel.text = "Weight: " + String(weightNum!)
```
avoid force unwrap with `if let`

splons commented on C7FIT/MapDetailView.swift:
```diff
@@ -0,0 +1,92 @@
+import UIKit
+import MapKit
+
+class MapDetailView: UIView, MKMapViewDelegate{
```
not typical for the view to be a delegate and does not look like this is being utilized?

splons commented on C7FIT/MapDetailViewController.swift:
```diff
@@ -0,0 +1,163 @@
+        return MKCoordinateRegion(
+            center: CLLocationCoordinate2D(latitude: (minLat + maxLat)/2, longitude: (minLng + maxLng)/2),
+            //change span doesnt work for long runs
+            span: MKCoordinateSpan(latitudeDelta: 0.01, longitudeDelta: 0.01))
```
you have the min/max lat/lng -- so why not setup the span from that info plus a little bit more (assuming you want a region that covers all of the points in the run?)

splons commented on C7FIT/MapView.swift:
```diff
@@ -0,0 +1,110 @@
+//        let centerTitleX = titleLabel.centerXAnchor.constraint(equalTo: centerXAnchor)
+//        let centerTitleY = titleLabel.centerYAnchor.constraint(equalTo: centerYAnchor)
+//        NSLayoutConstraint.activate([centerTitleX, centerTitleY])
+        let screenWidth = UIScreen.main.bounds.size.width
```
using screen width does not seem correct -- views should be constrained relative to their superview and peers

splons commented on C7FIT/MapViewController.swift:
```diff
@@ -0,0 +1,172 @@
+    
+    // MARK: Timer
+    func eachSecond(timer: Timer) {
+        seconds = seconds + 1
```
I don't think this will be very accurate, Timers are not guaranteed to fire accurately -- so this will drift

in startTrackRun capture a Date for the current time and use its timeIntervalSinceReferenceDate property against a current time to determine seconds

splons commented on C7FIT/MapViewController.swift:
```diff
@@ -0,0 +1,172 @@
+        let distanceQuantity = HKQuantity(unit: HKUnit.meter(), doubleValue: distance)
+        mapViewApp.distanceQuantity.text = "Distance: " + distanceQuantity.description
+        
+    // convert to mile per minute
```
does this code really convert to miles per minute? would not guess that by looking at it

splons commented on C7FIT/eBayDataManager.swift:
```diff
@@ -0,0 +1,89 @@
+                return
+            }
+            guard let dataJSON = try? JSONSerialization.jsonObject(with: data, options: []), let dataDict = dataJSON as? [String: Any] else {
+                return
```
okay to skip calling the completion block here?

splons commented on C7FIT/eBayDataManager.swift:
```diff
@@ -0,0 +1,89 @@
+                return
+            }
+            guard let dataJSON = try? JSONSerialization.jsonObject(with: data, options: []), let dataDict = dataJSON as? [String: Any] else {
+                return
```
same here, okay to not call the completion block?

splons commented on C7FIT/eBayItem.swift:
```diff
@@ -0,0 +1,52 @@
+ */
+struct eBayItem {
+    let mainImage: String?
+    var additionalImages = [String?]()
```
are these URLs? if so, either a better name for the var or maybe convert them to URL type

splons reviewed (commented):
Overall looks good. 

Spend time making the code consistent. Lots of small inconsistencies (always have a space following a colon, only use self. when needed, etc).

brandonlee503 committed:
Call completion in all cases

brandonlee503 committed:
Safely set image

brandonlee503 committed:
Updated eBayItem property names

Attempted to change to URL type, but it seemed a bit slower after
implementation. Maybe it’s just me 😅

brandonlee503 commented:
All of the code comments for store view have been addressed. The remaining comments pertain to the tools screen, which will be taken to account by @willmichael in his PR in the near future
